### ISSUE TITLE
Possible Bug: HypermediaSupportBeanDefinitionRegistrar forces / overwrites FAIL_ON_UNKNOWN_PROPERTIES to false

### ISSUE THREAD

**Issue by stefan-huettemann** [spring-projects/spring-hateoas#577]
We are using spring-boot 1.5.3.

[spring-hateoas-0.23.0.RELEASE-sources.jar!/org/springframework/hateoas/config/HypermediaSupportBeanDefinitionRegistrar.java:341](https://github.com/spring-projects/spring-hateoas/blob/0.23.0.RELEASE/src/main/java/org/springframework/hateoas/config/HypermediaSupportBeanDefinitionRegistrar.java#L341) forces `mapper.disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES
FAIL_ON_UNKNOWN_PROPERTIES);` on the ObjectMapper `HAL_OBJECT_MAPPER_BEAN_NAME`. 

We would require that our services responds with an error if a JSON (XML) with unknown properties is posted. We do not see that this would be in conflict with HATEOS principles. Using spring-boot we should be able to configure this by setting `spring.jackson.deserialization.fail-on-unknown-properties=true` in `application.properties` which is prevented by the code mentioned above.

It is unclear to me why this is done. I would consider this a bug. 

Note: this is related to issue #341.

Regards,
-Stefan

**Comment by olivergierke** [comment]
This is by intention and basically implements the [Robustness Principle](https://en.wikipedia.org/wiki/Robustness_principle).

**Comment by stefan-huettemann** [comment]
1. Why would HATEOS force my API not to fail on unknown properties? At least HATEOS should allow me (a simple / spring-boot like way) to opt out of this behavior.

1. In consequence this means that spring-boots auto-configuration is not working for `.jackson.deserialization.fail-on-unknown-properties=true` - which is (a very) unexpected and undocumented behavior.

1. I do not see how this implements robustness. If a client post/put/patches a payload that violates the API contract it should be notified about that. Else a client would receive a OK/200 even if the assumed change of state on the server has not taken place.

**Comment by olivergierke** [comment]
The reason is that a lot of the properties in `Resources` are mapped into a completely different format depending on the hypermedia type. If that format is then resubmitted, the deserialization would fail. I.e. without that enabled, very fundamental request, response exchanges would not work out of the box.

There a re a couple of things that need to be improved in terms of the interplay with Boot. Read more on this in related ticket in the Boot [issue tracker](https://github.com/spring-projects/spring-boot/issues/8174).

Receivers should be tolerant readers as otherwise every change to an API potentially becomes a breaking change. With REST, you want to avoid exactly that. Read more on that [here](https://martinfowler.com/bliki/TolerantReader.html).

**Comment by stefan-huettemann** [comment]
I totally agree on that disabling `fail-on-unknown-properties` is good for most cases and especially on public REST API.

I just do not see why this would be hardcoded and therefore forced upon APIs by spring-hateoas.

Do I understand correctly, that future releases of spring-hateos will eventually respect spring-boots `spring.jackson.deserialization.fail-on-unknown-properties=true` setting?

**Comment by stefan-huettemann** [comment]
In our (restricted) use case, we use a workaround to force `_halObjectMapper` to fail on unknown properties.

Could you please comment shortly if the following "workaround" code could be used; meaning that it will not break the *technical* aspects of spring-hateos working:
```
@Configuration
public class ObjectMapperConfiguration {

    @Value("${spring.jackson.deserialization.fail-on-unknown-properties:true}")
    private Boolean failOnUnknownProperties;

    private ApplicationContext applicationContext;

    @Autowired
    public ObjectMapperConfiguration(final ApplicationContext aApplicationContext) {
        applicationContext = aApplicationContext;
    }

    @Bean
    public ServletContextListener forceFailOnUnknownProperties() {

        return new ServletContextListener() {
            @Override
            public void contextInitialized(final ServletContextEvent sce) {
                for (final String theBeanName : applicationContext.getBeanNamesForType(ObjectMapper.class)) {
                    final ObjectMapper theBean = (ObjectMapper) applicationContext.getBean(theBeanName);
                    if (failOnUnknownProperties) {
                        theBean.enable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES);
                    } else {
                        theBean.disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES);
                    }
                }
            }

            @Override
            public void contextDestroyed(final ServletContextEvent sce) {

            }
        };
    }
}
```
Thanks & Regards,
-Stefan

**Comment by stefan-huettemann** [comment]
So I deleted my last comment since I see, that spring-hateoas does not "eat" what is produces. Meaning that by using `ResourceSupport`, `LinkBuilderSupport` and the like, a service will produce JSON/XML that will break corresponding consuming services if `fail-on-unknown-properties` is enabled.

May I suggest to put this (the enforcement of setting `fail-on-unknown-properties` to `false`) in a prominent place in the spring-hateos documentation? 

Thanks for your time.

**Comment by gregturn** [comment]
A basic point is that if you POST a HAL document, _links would break deserialization. 

As domain objects acquire new attributes, ignoring these attributes in older clients decreases friction and brittleness compared to RPC tool kits like CORBA and SOAP. 

```
{
   firstName: "Frodo",
   lastName: "Baggins"
   fullName: "Frodo Baggins"
}
```

A document like this, where fullName is a getter only, added in a recent release, and generated purely from first and last names, shouldn't break older clients. Hence, increased robustness and less desire for API versioning.

**Comment by stefan-huettemann** [comment]
I understand the benefits with respect to API changes. (Even though in our case we think that it is better if the server answers with an defined error message in the case when a posted property is not processed by the server - with all consequences that come with it).

I was not aware, that spring-hateoas does not respect the spring-boot configuration settings - which was causing some irritation on my side.

Also new to me is, that spring-hateos does not consume the documents it creates (e.g. Link -> "templated") when `fail-on-unknown-properties` is set to `true`.

I would have helped me if this behavior would be documented.

**Comment by gregturn** [comment]
Can we close this issue in light of https://github.com/spring-projects/spring-hateoas-examples?

This repository provides several examples using Spring HATEOAS and addresses the robustness principle in more detail.

Cool?

**Comment by stefan-huettemann** [comment]
Cool. Thanks for the elaborate examples. I think they help a lot.

I also think the issue should be closed since it is not a bug.

Having said this ... I still have some worries/questions about the hard-coded `fail-on-unknown-properties` though. Tickets do not seem the right way to elaborate on this. Can you point me to some forum / gitter group etc. where a possible discussion could take place?

**Comment by gregturn** [comment]
There is a spring-data gitter channel (https://gitter.im/spring-projects/spring-data), but I don't hang out there much. 

tl;dr github+JIRA+Slack+Email+Twitter = too many channels!
